source/language/openpulse.rst

```diff
@@ -0,0 +1,316 @@
+out the state of that qubit.
+
+.. code-block:: none
+
+   // Get the default input channel for qubit 0
+   transmit(%0)
+   // Get the default output channel for qubit 1
+   receive(%1)
+   // Get an input channel named "measure" for qubit 1
+   // Could be used to specify the measurement stimulus channel
+   transmit("measure", %1)
+   // Get an output channel named "cr1" for qubit 0
+   receive("cr1", %0)
+   // Get the default output channel for qubits 0 and 1
+   receive(%0, %1)
+
+Frames
+------
+
+It turns out to be quite useful to keep track of a set of carrier signals
+throughout the execution of a program. These carrier signals are called "frames"
+and are defined by :math:$A*e^{i\left(2\pi f t + \theta\right)}$, where `f` is
+frequency, `theta` is phase, and `A` is a scaling factor.
+
+Frames are attached to channels by a 1:1 correspondence. Every frame must 
+specify a frequency before any pulses can be played on that channel. The special 
+``prelude`` defcal is a good place to place these instructions.
+
+.. code-block:: none
+
+   defcal prelude {
+      // Required: set the frequency to 5GHz
+      set_frequency transmit(%1), 5e9;
+      // Optional: set the starting phase (otherwise it would start at 0)
+      set_phase transmit(%1), pi;
+   }
+
+Frames track the set of (frequency, phase, scale) throughout the program and 
+are manipulated using the following `set_` and `shift_` commands. Then when
+a pulse is played the carrier signal of the frame is mixed with the pulse itself
+to create the final signal which will be applied on the channel.
+
+Frequency has no default since it must be set at the beginning. Phase defaults
+to 0 and is a value of type ``angle``. Scale defaults to 1.0 and is a value of
+type ``float``. The exact precision of these three parameters is hardware
+specific.
+
+frameof command
+~~~~~~~~~~~~~~~
+
+The ``frame`` type is a compile-time type represnting a frame. It is useful
+when playing a pulse using the frame of a different channel.
+
+A value of type ``frame`` is retrieved by calling the ``frameof`` command
+and passing in a channel.
+
+Frame manipulation
+~~~~~~~~~~~~~~~~~~
+
+The (frequency, phase, scale) of a frame can be manipulated throughout program
+execution with the following instructions. After a call to one of these
+instructions, future references to that frame will have the new value.
+
+All of the following instructions have the same form. The first parameter can 
+either be a ``frame`` type or a ``channel`` type (in which case the 
+corresponding frame will be used). The second parameter is the value to shift
+or set. It must be the appropriate type, ``float`` for frequency, ``angle`` for
+phase, and ``float`` for scale. Again, the exact precision of these calculations
+is hardware specific.
+
+The relevant instructions are:
+
+- ``set_frequency`` and ``shift_frequency``
+- ``set_phase`` and ``shift_phase``
+- ``set_scale`` and ``shift_scale``
```

Comment:
What about `swap_phases`?

Reply:
Does this seem like a rather specific case of a more common operation? What if the phase relationship required is not a direct swap but a linear combination? Why do I need to swap the phase between two channels rather than copy from one channel to another and store the target channels original phase in a temporary location?

Reply:
I typically don't want to just store the original phase, though, but to allow it to continue accumulating at its frame frequency while I operate on another frame. `swap_phases` has an easily articulated use case in SWAP and iSWAP gates.

Reply:
Take a look at the latest diff and let me know how awkward that is to do now.